### PR TITLE
(CGT-2417) Created Composite Predicate and Spec

### PR DESCRIPTION
**Pull request by euptonqa**

### CODE REVIEW

File: test/predicates/CompositePredicateSpec.scala [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+    def predicate(dataModel: Option[AuthorisationDataModel], nino: Option[Nino]) = new CompositePredicate(appConfig, mockedService(dataModel, nino))(postSignURI,
+      notAuthorisedRedirectURI,
+      ivUpliftURI,
+      twoFactorURI,
+      authorisationURI)(hc)
+
+    "return true for page visibility when all supplied predicates are given an AuthContext that passes their associated checks" in {
+      val authContext = TestUserBuilder.compositePredicateUserPass
+      val result = predicate(Some(authorisationDataModelFail), ninoFail)(authContext, fakeRequest)
```
Comment (TomDudley92) [inline test/predicates/CompositePredicateSpec.scala]: This predicate is wrong as it should fail but is being overwritten by the authContext being passed in. Can this be changed please.

File: test/predicates/CompositePredicateSpec.scala [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+      val pageVisibility = await(result)
+
+      pageVisibility.isVisible shouldBe true
+    }
+
+    "return false for page visibility when all supplied predicates are given an AuthContext that passes their associated checks" in {
+      val authContext = TestUserBuilder.compositePredicateUserFail
+      val result = predicate(Some(authorisationDataModelPass), Some(Nino(ninoPass)))(authContext, fakeRequest)
```
Comment (TomDudley92) [inline test/predicates/CompositePredicateSpec.scala]: Same as the previous comment about the other predicate.

File: test/predicates/CompositePredicateSpec.scala [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    "return true for page visibility when all supplied predicates are given an AuthContext that passes their associated checks" in {
+      val authContext = TestUserBuilder.compositePredicateUserPass
+      val result = predicate(Some(authorisationDataModelFail), ninoFail)(authContext, fakeRequest)
+
+      val pageVisibility = await(result)
+
+      pageVisibility.isVisible shouldBe true
+    }
+
+    "return false for page visibility when all supplied predicates are given an AuthContext that passes their associated checks" in {
```
Comment (TomDudley92) [inline test/predicates/CompositePredicateSpec.scala]: Can this be worded a bit more clearly please

File: test/predicates/CompositePredicateSpec.scala [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+import common.Constants.AffinityGroup
+import uk.gov.hmrc.play.http.HeaderCarrier
+
+import scala.concurrent.Future
+
+
+@Singleton
+class CompositePredicateSpec extends UnitSpec with WithFakeApplication with MockitoSugar{
+
+  def mockedService(response: Option[AuthorisationDataModel], nino: Option[Nino]): AuthorisationService = {
```
Comment (TomDudley92) [inline test/predicates/CompositePredicateSpec.scala]: I know this isn't your issue but can you remove the option[Nino] from this test and also the AuthorisationServiceSpec.

File: app/predicates/CompositePredicate.scala [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+  override def children: Seq[PageVisibilityPredicate] = Seq (
+    new IVUpliftPredicate(new URI(ivUpliftUrl)),
+    new LoginPredicate(new URI(applicationConfig.governmentGateway)),
+    new TwoFAPredicate(new URI(twoFactorUrl)),
+    new NINOPredicate(new URI(ivUpliftUrl)),
+    new AffinityGroupPredicate(authorisationService)(new URI(affinityGroup))(hc)
+  )
+
+  private val ivUpliftURI: URI =
+    new URI(s"${ivUpliftUrl}?origin=CGT&"  +
```
Comment (TomDudley92) [inline app/predicates/CompositePredicate.scala]: A bit picky but can you remove the braces around the ivUpliftUrl please.

File: test/predicates/CompositePredicateSpec.scala [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package predicates
+
+import java.net.URI
+
+import builders.TestUserBuilder
```
Comment (TomDudley92) [inline test/predicates/CompositePredicateSpec.scala]: Can you remove redundant imports please.

File: app/predicates/CompositePredicate.scala (hunk 5 above)
Comment (euptonqa) [inline app/predicates/CompositePredicate.scala]: Just realised I was referencing the wrong variable too

**Pull request merged**